<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/ugcqz6ohbvff804xp84y4kqnvvk3bq1g.png" width = 300, align = "center"></a>

<h1 align=center><font size = 5>Lab: Working with a real world data-set using SQL and Python</font></h1>

# Introduction

This notebook shows how to work with a real world dataset using SQL and Python. In this lab you will:
1. Understand the dataset for Chicago Public School level performance 
1. Store the dataset in an Db2 database on IBM Cloud instance
1. Retrieve metadata about tables and columns and query data from mixed case columns
1. Solve example problems to practice your SQL skills including using built-in database functions

## Chicago Public Schools - Progress Report Cards (2011-2012) 

The city of Chicago released a dataset showing all school level performance data used to create School Report Cards for the 2011-2012 school year. The dataset is available from the Chicago Data Portal: https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t

This dataset includes a large number of metrics. Start by familiarizing yourself with the types of metrics in the database: https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?download=true

__NOTE__: Do not download the dataset directly from City of Chicago portal. Instead download a more database friendly version from the link below.
Now download a static copy of this database and review some of its contents:
https://ibm.box.com/shared/static/f9gjvj1gjmxxzycdhplzt01qtz0s7ew7.csv



### Store the dataset in a Table
In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. To analyze the data using SQL, it first needs to be stored in the database.

While it is easier to read the dataset into a Pandas dataframe and then PERSIST it into the database as we saw in the previous lab, it results in mapping to default datatypes which may not be optimal for SQL querying. For example a long textual field may map to a CLOB instead of a VARCHAR. 

Therefore, __it is highly recommended to manually load the table using the database console LOAD tool, as indicated in Week 2 Lab 1 Part II__. The only difference with that lab is that in Step 5 of the instructions you will need to click on create "(+) New Table" and specify the name of the table you want to create and then click "Next". 

##### Now open the Db2 console, open the LOAD tool, Select / Drag the .CSV file for the CHICAGO PUBLIC SCHOOLS dataset and load the dataset into a new table called __SCHOOLS__.

<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/uc4xjh1uxcc78ks1i18v668simioz4es.jpg"></a>

### Connect to the database
Let us now load the ipython-sql  extension and establish a connection with the database

In [1]:
%load_ext sql

In [2]:
# Enter the connection string for your Db2 on Cloud database instance below
# %sql ibm_db_sa://my-username:my-password@my-hostname:my-port/my-db-name
%sql ibm_db_sa://hln09807:l1jd9mx3m4z3x6%5Eh@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB

'Connected: hln09807@BLUDB'

### Query the database system catalog to retrieve table metadata

##### You can verify that the table creation was successful by retrieving the list of all tables in your schema and checking whether the SCHOOLS table was created

In [12]:
# type in your query to retrieve list of all tables in the database for your db2 schema (username)
%sql select TABSCHEMA, TABNAME, CREATE_TIME from SYSCAT.TABLES \
      where TABSCHEMA not in ('SYSIBM', 'SYSCAT', 'SYSSTAT', 'SYSIBMADM', 'SYSTOOLS', 'SYSPUBLIC')

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[('HLN09807', 'EMPLOYEES', datetime.datetime(2020, 2, 13, 6, 50, 12, 665442)),
 ('HLN09807', 'JOB_HISTORY', datetime.datetime(2020, 2, 13, 6, 50, 12, 854032)),
 ('HLN09807', 'JOBS', datetime.datetime(2020, 2, 13, 6, 50, 13, 25821)),
 ('HLN09807', 'DEPARTMENTS', datetime.datetime(2020, 2, 13, 6, 50, 13, 189624)),
 ('HLN09807', 'LOCATIONS', datetime.datetime(2020, 2, 13, 6, 50, 13, 365786)),
 ('HLN09807', 'EXPLAIN_INSTANCE', datetime.datetime(2020, 2, 23, 18, 4, 31, 741050)),
 ('HLN09807', 'EXPLAIN_STATEMENT', datetime.datetime(2020, 2, 23, 18, 4, 31, 938667)),
 ('HLN09807', 'EXPLAIN_ARGUMENT', datetime.datetime(2020, 2, 23, 18, 4, 32, 256398)),
 ('HLN09807', 'EXPLAIN_OBJECT', datetime.datetime(2020, 2, 23, 18, 4, 32, 506283)),
 ('HLN09807', 'EXPLAIN_OPERATOR', datetime.datetime(2020, 2, 23, 18, 4, 32, 615681)),
 ('HLN09807', 'EXPLAIN_PREDICATE', datetime.datetime(2020, 2, 23, 18, 4, 33, 49181)),
 ('HLN09807', 'EXPLAIN_STREAM', datetime.datetime(2020, 2, 23, 18, 4, 33, 278434)),
 ('HLN09807', 'EXPLAIN_DIAGNOSTIC', datetime.datetime(2020, 2, 23, 18, 4, 33, 516042)),
 ('HLN09807', 'EXPLAIN_DIAGNOSTIC_DATA', datetime.datetime(2020, 2, 23, 18, 4, 33, 727917)),
 ('HLN09807', 'OBJECT_METRICS', datetime.datetime(2020, 2, 23, 18, 4, 33, 969637)),
 ('HLN09807', 'ADVISE_INSTANCE', datetime.datetime(2020, 2, 23, 18, 4, 34, 148681)),
 ('HLN09807', 'ADVISE_INDEX', datetime.datetime(2020, 2, 23, 18, 4, 34, 317117)),
 ('HLN09807', 'ADVISE_WORKLOAD', datetime.datetime(2020, 2, 23, 18, 4, 34, 790389)),
 ('HLN09807', 'ADVISE_MQT', datetime.datetime(2020, 2, 23, 18, 4, 34, 990140)),
 ('HLN09807', 'ADVISE_PARTITION', datetime.datetime(2020, 2, 23, 18, 4, 35, 209334)),
 ('HLN09807', 'ADVISE_TABLE', datetime.datetime(2020, 2, 23, 18, 4, 35, 441732)),
 ('HLN09807', 'EXPLAIN_ACTUALS', datetime.datetime(2020, 2, 23, 18, 4, 37, 129042)),
 ('HLN09807', 'INSTRUCTOR', datetime.datetime(2020, 2, 22, 18, 6, 5, 657260)),
 ('HLN09807', 'INTERNATIONAL_STUDENT_TEST_SCORES', datetime.datetime(2020, 2, 22, 18, 49, 53, 530489)),
 ('HLN09807', 'CHICAGO_SOCIOECONOMIC_DATA', datetime.datetime(2020, 2, 22, 19, 48, 49, 24152)),
 ('HLN09807', 'PETSALE', datetime.datetime(2020, 2, 15, 6, 16, 13, 323157)),
 ('HLN09807', 'CHICAGO_PRGM', datetime.datetime(2020, 2, 24, 3, 22, 17, 106084))]

Double-click __here__ for a hint

<!--
In Db2 the system catalog table called SYSCAT.TABLES contains the table metadata
-->

Double-click __here__ for the solution.

<!-- Solution:

%sql select TABSCHEMA, TABNAME, CREATE_TIME from SYSCAT.TABLES where TABSCHEMA='YOUR-DB2-USERNAME'

or, you can retrieve list of all tables where the schema name is not one of the system created ones:

%sql select TABSCHEMA, TABNAME, CREATE_TIME from SYSCAT.TABLES \
      where TABSCHEMA not in ('SYSIBM', 'SYSCAT', 'SYSSTAT', 'SYSIBMADM', 'SYSTOOLS', 'SYSPUBLIC')
      
or, just query for a specifc table that you want to verify exists in the database
%sql select * from SYSCAT.TABLES where TABNAME = 'SCHOOLS'

-->

### Query the database system catalog to retrieve column metadata

##### The SCHOOLS table contains a large number of columns. How many columns does this table have?

In [18]:
# type in your query to retrieve the number of columns in the SCHOOLS table
%sql select * from syscat.columns where tabname = 'CHICAGO_PRGM'

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[('HLN09807', 'CHICAGO_PRGM', 'CPS_Performance_Policy_Level', 14, 'SYSIBM  ', 'VARCHAR', 15, 0, 'OCTETS', 15, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 5, "'NDA'", "'Level 2'", 12, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 7, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM', 'HEALTHY_SCHOOL_CERTIFIED', 15, 'SYSIBM  ', 'VARCHAR', 3, 0, 'OCTETS', 3, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 2, "'Yes'", "'No'", 7, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 2, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM', 'Safety_Icon', 16, 'SYSIBM  ', 'VARCHAR', 11, 0, 'OCTETS', 11, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 6, "'Very Weak'", "'NDA'", 11, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 6, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM', 'SAFETY_SCORE', 17, 'SYSIBM  ', 'SMALLINT', 2, 0, None, None, None, 'Y', 0, None, None, ' ', ' ', 86, '99', '5', 3, None, 0, 20, 10, 53, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, -1, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM', 'Family_Involvement_Icon', 18, 'SYSIBM  ', 'VARCHAR', 11, 0, 'OCTETS', 11, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 6, "'Very Weak'", "'NDA'", 10, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 5, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM', 'Family_Involvement_Score', 19, 'SYSIBM  ', 'VARCHAR', 3, 0, 'OCTETS', 3, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 72, "'99'", "'17'", 7, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 2, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM', 'Environment_Icon', 20, 'SYSIBM  ', 'VARCHAR', 11, 0, 'OCTETS', 11, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 6, "'Very Weak'", "'NDA'", 11, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 6, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM', 'Environment_Score', 21, 'SYSIBM  ', 'SMALLINT', 2, 0, None, None, None, 'Y', 0, None, None, ' ', ' ', 80, '99', '2', 3, None, 0, 20, 10, 53, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, -1, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM', 'Instruction_Icon', 22, 'SYSIBM  ', 'VARCHAR', 11, 0, 'OCTETS', 11, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 6, "'Very Weak'", "'NDA'", 11, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 6, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM', 'Instruction_Score', 23, 'SYSIBM  ', 'SMALLINT', 2, 0, None, None, None, 'Y', 0, None, None, ' ', ' ', 83, '99', '11', 3, None, 0, 20, 10, 53, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, -1, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM', 'Leaders_Icon', 24, 'SYSIBM  ', 'VARCHAR', 11, 0, 'OCTETS', 11, None, 'Y', 1208, 'SYSIBM', 'IDENTITY', ' ', ' ', 6, "'Very Weak'", "'NDA'", 10, None, 0, 20, 10, 0, None, None, None, None, None, None, None, None, ' ', 0, -1, 'N', 'N', ' ', None, 'O', None, -1.0, -1, -1, 5, None, None, 'N', 'N', 'N', -1, -1.0, None, None, 'N', None),
 ('HLN09807', 'CHICAGO_PRGM

Double-click __here__ for a hint

<!--
In Db2 the system catalog table called SYSCAT.COLUMNS contains the column metadata
-->

Double-click __here__ for the solution.

<!-- Solution:

%sql select count(*) from SYSCAT.COLUMNS where TABNAME = 'SCHOOLS'

-->

Now retrieve the the list of columns in SCHOOLS table and their column type (datatype) and length.

In [71]:
# type in your query to retrieve all column names in the SCHOOLS table along with their datatypes and length
%sql select COLNAME, TYPENAME, LENGTH from SYSCAT.COLUMNS where TABNAME = 'CHICAGO_PRGM'

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[('CPS_Performance_Policy_Level', 'VARCHAR', 15),
 ('HEALTHY_SCHOOL_CERTIFIED', 'VARCHAR', 3),
 ('Safety_Icon', 'VARCHAR', 11),
 ('SAFETY_SCORE', 'SMALLINT', 2),
 ('Family_Involvement_Icon', 'VARCHAR', 11),
 ('Family_Involvement_Score', 'VARCHAR', 3),
 ('Environment_Icon', 'VARCHAR', 11),
 ('Environment_Score', 'SMALLINT', 2),
 ('Instruction_Icon', 'VARCHAR', 11),
 ('Instruction_Score', 'SMALLINT', 2),
 ('Leaders_Icon', 'VARCHAR', 11),
 ('Leaders_Score', 'VARCHAR', 3),
 ('Teachers_Icon', 'VARCHAR', 11),
 ('Teachers_Score', 'VARCHAR', 3),
 ('Parent_Engagement_Icon', 'VARCHAR', 7),
 ('Parent_Engagement_Score', 'VARCHAR', 3),
 ('Parent_Environment_Icon', 'VARCHAR', 7),
 ('Parent_Environment_Score', 'VARCHAR', 3),
 ('AVERAGE_STUDENT_ATTENDANCE', 'VARCHAR', 6),
 ('Rate_of_Misconducts__per_100_students_', 'DECIMAL', 5),
 ('Average_Teacher_Attendance', 'VARCHAR', 6),
 ('Individualized_Education_Program_Compliance_Rate', 'VARCHAR', 7),
 ('Pk_2_Literacy__', 'VARCHAR', 4),
 ('Pk_2_Math__', 'VARCHAR', 4),
 ('Gr3_5_Grade_Level_Math__', 'VARCHAR', 4),
 ('Gr3_5_Grade_Level_Read__', 'VARCHAR', 4),
 ('School_ID', 'INTEGER', 4),
 ('NAME_OF_SCHOOL', 'VARCHAR', 65),
 ('Elementary, Middle, or High School', 'VARCHAR', 2),
 ('Street_Address', 'VARCHAR', 30),
 ('City', 'VARCHAR', 7),
 ('State', 'VARCHAR', 2),
 ('ZIP_Code', 'INTEGER', 4),
 ('Phone_Number', 'VARCHAR', 14),
 ('Link', 'VARCHAR', 78),
 ('Network_Manager', 'VARCHAR', 40),
 ('Collaborative_Name', 'VARCHAR', 34),
 ('Adequate_Yearly_Progress_Made_', 'VARCHAR', 3),
 ('Track_Schedule', 'VARCHAR', 12),
 ('CPS_Performance_Policy_Status', 'VARCHAR', 16),
 ('Gr3_5_Keep_Pace_Read__', 'VARCHAR', 4),
 ('Gr3_5_Keep_Pace_Math__', 'VARCHAR', 4),
 ('Gr6_8_Grade_Level_Math__', 'VARCHAR', 4),
 ('Gr6_8_Grade_Level_Read__', 'VARCHAR', 4),
 ('Gr6_8_Keep_Pace_Math_', 'VARCHAR', 4),
 ('Gr6_8_Keep_Pace_Read__', 'VARCHAR', 4),
 ('Gr_8_Explore_Math__', 'VARCHAR', 4),
 ('Gr_8_Explore_Read__', 'VARCHAR', 4),
 ('ISAT_Exceeding_Math__', 'DECIMAL', 4),
 ('ISAT_Exceeding_Reading__', 'DECIMAL', 4),
 ('ISAT_Value_Add_Math', 'DECIMAL', 3),
 ('ISAT_Value_Add_Read', 'DECIMAL', 3),
 ('ISAT_Value_Add_Color_Math', 'VARCHAR', 6),
 ('ISAT_Value_Add_Color_Read', 'VARCHAR', 6),
 ('Students_Taking__Algebra__', 'VARCHAR', 4),
 ('Students_Passing__Algebra__', 'VARCHAR', 4),
 ('9th Grade EXPLORE (2009)', 'VARCHAR', 4),
 ('9th Grade EXPLORE (2010)', 'VARCHAR', 4),
 ('10th Grade PLAN (2009)', 'VARCHAR', 4),
 ('10th Grade PLAN (2010)', 'VARCHAR', 4),
 ('Net_Change_EXPLORE_and_PLAN', 'VARCHAR', 3),
 ('11th Grade Average ACT (2011)', 'VARCHAR', 4),
 ('Net_Change_PLAN_and_ACT', 'VARCHAR', 3),
 ('College_Eligibility__', 'VARCHAR', 4),
 ('Graduation_Rate__', 'VARCHAR', 4),
 ('College_Enrollment_Rate__', 'VARCHAR', 4),
 ('COLLEGE_ENROLLMENT', 'SMALLINT', 2),
 ('General_Services_Route', 'SMALLINT', 2),
 ('Freshman_on_Track_Rate__', 'VARCHAR', 4),
 ('X_COORDINATE', 'DECIMAL', 13),
 ('Y_COORDINATE', 'DECIMAL', 13),
 ('Latitude', 'DECIMAL', 18),
 ('Longitude', 'DECIMAL', 18),
 ('COMMUNITY_AREA_NUMBER', 'SMALLINT', 2),
 ('COMMUNITY_AREA_NAME', 'VARCHAR', 22),
 ('Ward', 'SMALLINT', 2),
 ('Police_District', 'SMALLINT', 2),
 ('Location', 'VARCHAR', 27)]

Double-click __here__ for the solution.

<!-- Solution:

%sql select COLNAME, TYPENAME, LENGTH from SYSCAT.COLUMNS where TABNAME = 'SCHOOLS'

or

%sql select distinct(NAME), COLTYPE, LENGTH from SYSIBM.SYSCOLUMNS where TBNAME = 'SCHOOLS'

-->

### Questions
1. Is the column name for the "SCHOOL ID" attribute in upper or mixed case?
1. What is the name of "Community Area Name" column in your table? Does it have spaces?
1. Are there any columns in whose names the spaces and paranthesis (round brackets) have been replaced by the underscore character "_"?

## Problems

### Problem 1

##### How many Elementary Schools are in the dataset?

In [20]:
%sql select count(*) from CHICAGO_PRGM where "Elementary, Middle, or High School" = 'ES'

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[(Decimal('462'),)]

Double-click __here__ for a hint

<!--
Which column specifies the school type e.g. 'ES', 'MS', 'HS'?
-->

Double-click __here__ for another hint

<!--
Does the column name have mixed case, spaces or other special characters?
If so, ensure you use double quotes around the "Name of the Column"
-->

Double-click __here__ for the solution.

<!-- Solution:

%sql select count(*) from SCHOOLS where "Elementary, Middle, or High School" = 'ES'

Correct answer: 462

-->

### Problem 2

##### What is the highest Safety Score?

In [21]:
%sql select max(SAFETY_SCORE) from CHICAGO_PRGM

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[(99,)]

Double-click __here__ for a hint

<!--
Use the MAX() function
-->

Double-click __here__ for the solution.

<!-- Hint:

%sql select MAX(Safety_Score) AS MAX_SAFETY_SCORE from SCHOOLS

Correct answer: 99
-->


### Problem 3

##### Which schools have highest Safety Score?

In [28]:
%%sql
select NAME_OF_SCHOOL , Safety_Score from CHICAGO_PRGM WHERE 
Safety_Score = (select MAX(Safety_Score) from CHICAGO_PRGM)

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[('Abraham Lincoln Elementary School', 99),
 ('Alexander Graham Bell Elementary School', 99),
 ('Annie Keller Elementary Gifted Magnet School', 99),
 ('Augustus H Burley Elementary School', 99),
 ('Edgar Allan Poe Elementary Classical School', 99),
 ('Edgebrook Elementary School', 99),
 ('Ellen Mitchell Elementary School', 99),
 ('James E McDade Elementary Classical School', 99),
 ('James G Blaine Elementary School', 99),
 ('LaSalle Elementary Language Academy', 99),
 ('Mary E Courtenay Elementary Language Arts Center', 99),
 ('Northside College Preparatory High School', 99),
 ('Northside Learning Center High School', 99),
 ('Norwood Park Elementary School', 99),
 ('Oriole Park Elementary School', 99),
 ('Sauganash Elementary School', 99),
 ('Stephen Decatur Classical Elementary School', 99),
 ('Talman Elementary School', 99),
 ('Wildwood Elementary School', 99)]

Double-click __here__ for the solution.

<!-- Solution:
In the previous problem we found out that the highest Safety Score is 99, so we can use that as an input in the where clause:

%sql select Name_of_School, Safety_Score from SCHOOLS where Safety_Score = 99

or, a better way:

%sql select Name_of_School, Safety_Score from SCHOOLS where \
  Safety_Score= (select MAX(Safety_Score) from SCHOOLS)


Correct answer: several schools with with Safety Score of 99.
-->


### Problem 4

##### What are the top 10 schools with the highest "Average Student Attendance"?


In [35]:
%%sql
select NAME_OF_SCHOOL , AVERAGE_STUDENT_ATTENDANCE from CHICAGO_PRGM order by 
AVERAGE_STUDENT_ATTENDANCE desc nulls last limit 10 

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[('John Charles Haines Elementary School', '98.40%'),
 ('James Ward Elementary School', '97.80%'),
 ('Edgar Allan Poe Elementary Classical School', '97.60%'),
 ('Orozco Fine Arts & Sciences Elementary School', '97.60%'),
 ('Rachel Carson Elementary School', '97.60%'),
 ('Annie Keller Elementary Gifted Magnet School', '97.50%'),
 ('Andrew Jackson Elementary Language Academy', '97.40%'),
 ('Lenart Elementary Regional Gifted Center', '97.40%'),
 ('Disney II Magnet School', '97.30%'),
 ('John H Vanderpoel Elementary Magnet School', '97.20%')]

Double-click __here__ for the solution.

<!-- Solution:

%sql select Name_of_School, Average_Student_Attendance from SCHOOLS \
    order by Average_Student_Attendance desc nulls last limit 10 

-->

### Problem 5

##### Retrieve the list of 5 Schools with the lowest Average Student Attendance sorted in ascending order based on attendance

In [42]:
%%sql
select NAME_OF_SCHOOL , AVERAGE_STUDENT_ATTENDANCE from CHICAGO_PRGM order by 
AVERAGE_STUDENT_ATTENDANCE asc nulls last limit 5

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[('Richard T Crane Technical Preparatory High School', '57.90%'),
 ('Barbara Vick Early Childhood & Family Center', '60.90%'),
 ('Dyett High School', '62.50%'),
 ('Wendell Phillips Academy High School', '63.00%'),
 ('Orr Academy High School', '66.30%')]

Double-click __here__ for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from SCHOOLS \
     order by Average_Student_Attendance \
     fetch first 5 rows only

-->


### Problem 6

##### Now remove the '%' sign from the above result set for Average Student Attendance column

In [44]:
%sql select NAME_OF_SCHOOL , REPLACE(AVERAGE_STUDENT_ATTENDANCE , '%' , '') \
from CHICAGO_PRGM order by \
AVERAGE_STUDENT_ATTENDANCE asc nulls last limit 5


 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[('Richard T Crane Technical Preparatory High School', '57.90'),
 ('Barbara Vick Early Childhood & Family Center', '60.90'),
 ('Dyett High School', '62.50'),
 ('Wendell Phillips Academy High School', '63.00'),
 ('Orr Academy High School', '66.30')]

Double-click __here__ for a hint

<!--
Use the REPLACE() function to replace '%' with ''
See documentation for this function at:
https://www.ibm.com/support/knowledgecenter/en/SSEPGG_10.5.0/com.ibm.db2.luw.sql.ref.doc/doc/r0000843.html
-->

Double-click __here__ for the solution.

<!-- Hint:

%sql SELECT Name_of_School, REPLACE(Average_Student_Attendance, '%', '') \
     from SCHOOLS \
     order by Average_Student_Attendance \
     fetch first 5 rows only

-->


### Problem 7

##### Which Schools have Average Student Attendance lower than 70%?

In [47]:
%sql select NAME_OF_SCHOOL , AVERAGE_STUDENT_ATTENDANCE \
from CHICAGO_PRGM WHERE \
CAST(REPLACE(AVERAGE_STUDENT_ATTENDANCE,'%','') AS DOUBLE) < 70

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[('Barbara Vick Early Childhood & Family Center', '60.90%'),
 ('Chicago Vocational Career Academy High School', '68.80%'),
 ('Dyett High School', '62.50%'),
 ('Manley Career Academy High School', '66.80%'),
 ('Orr Academy High School', '66.30%'),
 ('Richard T Crane Technical Preparatory High School', '57.90%'),
 ('Roberto Clemente Community Academy High School', '69.60%'),
 ('Wendell Phillips Academy High School', '63.00%')]

Double-click __here__ for a hint

<!--
The datatype of the "Average_Student_Attendance" column is varchar.
So you cannot use it as is in the where clause for a numeric comparison.
First use the CAST() function to cast it as a DECIMAL or DOUBLE
e.g. CAST("Column_Name" as DOUBLE)
or simply: DECIMAL("Column_Name")
-->

Double-click __here__ for another hint

<!--
Don't forget the '%' age sign needs to be removed before casting
-->

Double-click __here__ for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from SCHOOLS \
     where CAST ( REPLACE(Average_Student_Attendance, '%', '') AS DOUBLE ) < 70 \
     order by Average_Student_Attendance
     
or,

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from SCHOOLS \
     where DECIMAL ( REPLACE(Average_Student_Attendance, '%', '') ) < 70 \
     order by Average_Student_Attendance

-->


### Problem 8

##### Get the total College Enrollment for each Community Area

In [97]:
%sql SELECT SUM(College_Enrollment) , COMMUNITY_AREA_NUMBER FROM CHICAGO_PRGM \
group by COMMUNITY_AREA_NUMBER 


 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
(ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/LINUXX8664] SQL0119N  An expression starting with "COLLEGE_ENROLLMENT" specified in a SELECT clause, HAVING clause, or ORDER BY clause is not specified in the GROUP BY clause or it is in a SELECT clause, HAVING clause, or ORDER BY clause with a column function and no GROUP BY clause is specified.  SQLSTATE=42803 SQLCODE=-119
[SQL: select Max(t) from ( SELECT College_Enrollment as t, COMMUNITY_AREA_NUMBER FROM CHICAGO_PRGM  group by COMMUNITY_AREA_NUMBER )]
(Background on this error at: http://sqlalche.me/e/f405)


Double-click __here__ for a hint

<!--
Verify the exact name of the Enrollment column in the database
Use the SUM() function to add up the Enrollments for each Community Area
-->

Double-click __here__ for another hint

<!--
Don't forget to group by the Community Area
-->

Double-click __here__ for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from SCHOOLS \
   group by Community_Area_Name 

-->


### Problem 9

##### Get the 5 Community Areas with the least total College Enrollment  sorted in ascending order 

In [70]:
%sql  SELECT SUM(College_Enrollment) AS TOTAL_College_Enrollment, Community_Area_Name FROM CHICAGO_PRGM \
group by Community_Area_Name  ORDER BY TOTAL_College_Enrollment DESC LIMIT 5


 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
(ibm_db_dbi.ProgrammingError) ibm_db_dbi::ProgrammingError: SQLNumResultCols failed: [IBM][CLI Driver][DB2/LINUXX8664] SQL0104N  An unexpected token "WHERE" was found following "Community_Area_Name".  Expected tokens may include:  "NPS_DOUBLE_DOT".  SQLSTATE=42601 SQLCODE=-104
[SQL: SELECT SUM(College_Enrollment) AS TOTAL_College_Enrollment, Community_Area_Name FROM CHICAGO_PRGM  group by Community_Area_Name  WHERE  TOTAL_College_Enrollment = 4638]
(Background on this error at: http://sqlalche.me/e/f405)


Double-click __here__ for a hint

<!--
Order the previous query and limit the number of rows you fetch
-->

Double-click __here__ for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from SCHOOLS \
   group by Community_Area_Name \
   order by TOTAL_ENROLLMENT asc \
   fetch first 5 rows only

-->

### Problem 10

##### Get the hardship index for the community area which has College Enrollment of 4638

In [88]:
%%sql 
select hardship_index 
   from chicago_socioeconomic_data CD, CHICAGO_PRGM CPS where CD.ca = CPS.community_area_number 
      and college_enrollment = 4368

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[(6.0,)]

Double-click __here__ for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%%sql 
select hardship_index 
   from chicago_socioeconomic_data CD, schools CPS 
   where CD.ca = CPS.community_area_number 
      and college_enrollment = 4368

-->

### Problem 11

##### Get the hardship index for the community area which has the highest value for College Enrollment

In [100]:
%%sql 
select ca , community_area_name , hardship_index from chicago_socioeconomic_data  where 
ca in (select community_area_number from CHICAGO_PRGM order by college_enrollment desc limit 1)

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[(5.0, 'North Center', 6.0)]

In [91]:
%sql select ca, community_area_name, hardship_index from chicago_socioeconomic_data \
   where ca in \
   ( select community_area_number from CHICAGO_PRGM order by college_enrollment desc limit 1 )

 * ibm_db_sa://hln09807:***@dashdb-txn-sbox-yp-dal09-03.services.dal.bluemix.net:50000/BLUDB
Done.


[(5.0, 'North Center', 6.0)]

Double-click __here__ for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%sql select ca, community_area_name, hardship_index from chicago_socioeconomic_data \
   where ca in \
   ( select community_area_number from schools order by college_enrollment desc limit 1 )

-->

## Summary

##### In this lab you learned how to work with a real word dataset using SQL and Python. You learned how to query columns with spaces or special characters in their names and with mixed case names. You also used built in database functions and practiced how to sort, limit, and order result sets, as well as used sub-queries and worked with multiple tables.

Copyright &copy; 2018 [cognitiveclass.ai](cognitiveclass.ai?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).
